tempesta_fw/tls.c

```diff
@@ -133,6 +135,20 @@ tfw_tls_msg_process(void *conn, TfwFsmData *data)
 		       parsed, skb->len);
 	}
 
+	if (msglen > 0) {
+		/*
+		 * We were in the middle of the TLS record processing, when
+		 * an skb arrived, so there is actually a saved msglen field
+		 * value which we can use.
+		 */
+		data_up.len = msglen - ttls_xfrm_taglen(&tls->xfrm) -
+			      ttls_expiv_len(&tls->xfrm);
+	} else {
+		WARN_ON(tls->io_in.msglen != 0);
+		data_up.len = parsed - ttls_xfrm_taglen(&tls->xfrm) -
+			      ttls_payload_off(&tls->xfrm);
+	}
```

Comment:
The parsing code eventually called by `ttls_recv()` resets the `tls->io_in` by zeroing everything past `__initoff` after the whole record is parsed. So there are two cases.
First: TLS record was spanning over multiple skbs. We just finished processing a last skb with the record. At this moment `tls->io_in` was already reset, but it wasn't cleared before we called `ss_skb_process()`, so what we saved to `msglen` was the current record length.
Second: TLS record was entirely in the skb. Then `msglen` will be `0` since previous record was processed and `tls->io_in` was reset previously. Then we processed the current record, then `tls->io_in` was reset again. So there is no way to get record length from `tls->io_in`. But fortunately, we have `parsed`, which we use to infer the current record length.

This is a shady part in the patch. It's far from obvious. The possible solution was to change `tls->io_in` processing to clear it before parsing, not after. But that was going to complicate the patch more, so I stopped there.

Reply:
OK, I see.

This code takes control only for application data records and it'd be good to have appropriate `WARN_ON()` by the end of the `switch` statement.

I'm still not sure about the `else` branch: if we have 2 chunks of data - e.g. 4 bytes of the header and the rest of the record - `msglen` on receiving the second chunk is still 0 since we didn't read the header fully, but `parsed` contains only length of the second chunk and the first 4 bytes seem just lost in the computation.

Reply:
> if we have 2 chunks of data - e.g. 4 bytes of the header and the rest of the record

You are right, that's the bug. Thanks!

Reply:
I've moved `msglen` field before `__initoff` so it survives "the purge", and can be used without tricks.